src/details/ArborX_Ray.hpp

```diff
@@ -198,6 +199,231 @@ bool intersects(Ray const &ray, Box const &box)
+                                             float &t)
+{
+  float x3 = edge_vertex_1[0];
+  float y3 = edge_vertex_1[1];
+  float x4 = edge_vertex_2[0];
+  float y4 = edge_vertex_2[1];
+
+  float y2 = std::fabs(y3) > std::fabs(y4) ? y3 : y4;
+
+  float det = y2 * (x3 - x4);
+
+  //  the ray is parallel to the edge if det == 0.0
+  //  When the ray overlaps the edge (x3==x4==0.0), it also returns false,
+  //  and the intersection will be captured by the other two edges.
+  if (det == 0)
+  {
+    return false;
+  }
+  t = (x3 * y4 - x4 * y3) / det * y2;
+
+  float u = x3 * y2 / det;
+
+  auto const epsilon = 0.00001f;
+  return (u >= 0 - epsilon && u <= 1 + epsilon);
+}
+
+// The algorithm is described in
+// Watertight Ray/Triangle Intersection
+// [1] Woop, S. et al. (2013),
+// Journal of Computer Graphics Techniques Vol. 2(1)
+// The major difference is that here we return the intersection points
+// when the ray and the triangle is coplanar.
+// In the paper, they just need the boolean return.
+KOKKOS_INLINE_FUNCTION
+bool intersection(Ray const &ray, Triangle const &triangle, float &tmin,
+                  float &tmax)
+{
+  auto dir = ray.direction();
+  // normalize the direction vector by its largest component.
+  auto kz = findLargestComp(dir);
+  int kx = (kz + 1) % 3;
+  int ky = (kz + 2) % 3;
+
+  if (dir[kz] < 0)
+    KokkosExt::swap(kx, ky);
+
+  Vector s;
+
+  s[2] = 1.0f / dir[kz];
+  s[0] = dir[kx] * s[2];
+  s[1] = dir[ky] * s[2];
+
+  // calculate vertices relative to ray origin
+  Vector const oA = makeVector(ray.origin(), triangle.a);
+  Vector const oB = makeVector(ray.origin(), triangle.b);
+  Vector const oC = makeVector(ray.origin(), triangle.c);
+
+  Point A;
+  Point B;
+  Point C;
+
+  // perform shear and scale of vertices
+  A[0] = oA[kx] - s[0] * oA[kz];
+  A[1] = oA[ky] - s[1] * oA[kz];
+  B[0] = oB[kx] - s[0] * oB[kz];
+  B[1] = oB[ky] - s[1] * oB[kz];
+  C[0] = oC[kx] - s[0] * oC[kz];
+  C[1] = oC[ky] - s[1] * oC[kz];
+
+  // calculate scaled barycentric coordinates
+  float u = C[0] * B[1] - C[1] * B[0];
+  float v = A[0] * C[1] - A[1] * C[0];
+  float w = B[0] * A[1] - B[1] * A[0];
+
+  // fallback to edge test using double precision
+  if (u == 0 || v == 0 || w == 0)
+  {
+    u = (double)C[0] * B[1] - (double)C[1] * B[0];
+    v = (double)A[0] * C[1] - (double)A[1] * C[0];
+    w = (double)B[0] * A[1] - (double)B[1] * A[0];
+  }
+
+  constexpr auto inf = KokkosExt::ArithmeticTraits::infinity<float>::value;
+  tmin = inf;
+  tmax = -inf;
+
+  // 'Back-face culling' is not supported.
+  // Back-facing culling is to check whether
+  // a surface is 'visible' to a ray, which requires
+  // consistent definition of the facing of triangles.
+  // Once the facing of triangle is defined,
+  // only one of the conditions is needed,
+  // either (u < 0 || v < 0 || w < 0) or
+  // (u > 0 || v > 0 || w > 0), for Back-facing culling.
+  auto const epsilon = 0.00001f;
```

Comment:
Why do you have an epsilon?  The original code does not.

Reply:
the value of u, v, w will scale with the size of the triangle.
I am still figuring out how to normalize them so that the u, v, w won't change for the test I added in the latest commit (line 533 in tstRay.cpp).

Reply:
Can't we normalize coordinates by the maximum of the difference between vertices and ray origin and rescale the ray length to be one?